nanoFramework.System.Net/NetworkHelper/NetworkHelper.cs

```diff
@@ -0,0 +1,300 @@
+            // loop until cancellation token expires or there is an IP address
+            while (!token.IsCancellationRequested && !CheckIP())
+            {
+                Thread.Sleep(200);
+            }
+
+            // handle cancellation token expiration
+            if (token.IsCancellationRequested)
+            {
+                // update status
+                _networkHelperStatus = NetworkHelperStatus.TokenExpiredWaitingIPAddress;
+
+                // operation failed
+                return false;
+            }
+
+            // if DateTime was requested, go for it
+            if (requiresDateTime)
+            {
+                Debug.WriteLine("Waiting for valid DateTime system clock...");
+
+                while (!token.IsCancellationRequested && (DateTime.UtcNow.Year < 2021))
+                {
+                    Thread.Sleep(200);
+                }
+            }
+
+            // handle cancellation token expiration
+            if (token.IsCancellationRequested)
+            {
+                // update status
+                _networkHelperStatus = NetworkHelperStatus.TokenExpiredWaitingDateTime;
+
+                // operation failed
+                return false;
+            }
+            else
+            {
+                Debug.WriteLine($"We have a valid date {DateTime.UtcNow}");
+            }
+
+            // update status
+            _networkHelperStatus = NetworkHelperStatus.NetworkIsReady;
+
+            // operation succeeded
+            return true;
+        }
+
+        private static void WorkingThread()
+        {
+            // check if we have an IP
+            CheckIP();
+
+            // wait here until we have an IP address
+            _ipAddressAvailable.WaitOne();
+
+            if (_requiresDateTime)
+            {
+                // wait until there is a valid DateTime
+                WaitForValidDateTime();
+            }
+            else
+            {
+                // all conditions met
+                _networkReady.Set();
+            }
+
+            // update flag
+            _networkHelperStatus = NetworkHelperStatus.NetworkIsReady;
+        }
+
+        /// <summary>
+        /// Checks if there is a valid DateTime and sets the <see cref="_ipAddressAvailable"/> event when that happens, if ever.
+        /// </summary>
+        private static void WaitForValidDateTime()
+        {
+            // if SNTP is available and enabled on target device this can be skipped because we should have a valid date & time
+            while (DateTime.UtcNow.Year < 2021)
+            {
+                Debug.WriteLine("Checking for valid date time...");
+
+                // wait for valid date & time
+                Thread.Sleep(1000);
+            }
+        }
+
+        /// <summary>
+        /// Checks if there is an IPAddress assigned to the device and sets the <see cref="_ipAddressAvailable"/> event accordingly.
+        /// </summary>
+        private static bool CheckIP()
```

Comment:
you should pass the type of interface you want to check. Reason is: I can setup a valid IP address on my wifi interface. But on the ethernet interface, I'm waiting for a valid IP Address. And in that case, it will tell me I have a valid IP Address. Which is not what I can expect.

Reply:
There will be a Network help class in the WiFi library "extending" this one with the appropriate methods (basically what's in the existing network helper). It will be able to tap into this one after the wifi stuff is taken care of.